# Single cell analysis on CLL data
This repository contains all the work did during an internship in CRCT about the single-cell analysis on Chronic Lymphocytic Leukemia (CLL) data. 
The first aim is to construct a solid cell type annotation.
1- Pre-precessing (QC, normalization, scaling, filtering, doublet and non coding genes elimination...)
2- Annotation with differents tools, manually and automaticly

## Dataset 
The data used in this study comes from autologous cell cultures. These cultures were generated from blood samples of 3 patients with chronic lymphocytic leukemia (CLL). They contain peripheral blood mononuclear cells (PBMCs) with the addition of exogenous neutrophils. Samples were collected at five specific time points: Day 1, Day 4, Day 8, Day 11, and Day 14, to monitor this progression. Cells from each time point were analyzed using CITE-seq (RNA + ADT). 
The alignment and quantification pipeline was not performed as part of this work. This analysis starts directly from the gene count matrices generated by 10X Genomics.
Specifically, the input data consist of the filtered_feature_bc_matrix.h5 and raw_feature_bc_matrix.h5 files, which correspond to the processed outputs of the Cell Ranger pipeline.

## Author 
Augustine Blanc-Boekholt 
