# Single cell analysis on CLL data

This repository contains all the work carried out during an internship at the CRCT, focusing on single-cell RNA sequencing (scRNA-seq) analysis of Chronic Lymphocytic Leukemia (CLL) data.
The primary objective of this project was to establish a robust and reliable cell type annotation. The workflow includes the following main steps:

- Pre-processing of the scRNA-seq data, including: Quality control (QC), Filtering, Normalization and scaling, Removal of doublets, Exclusion of non-coding genes

+ Cell type annotation, performed using: Manual annotation based on marker genes, Automated annotation tools (SingleR, scPred, Azimuth...)

* Cross-validation between methods to ensure consistency and accuracy

## Dataset 
The data used in this study comes from autologous cell cultures. These cultures were generated from blood samples of 3 patients with chronic lymphocytic leukemia (CLL). They contain peripheral blood mononuclear cells (PBMCs) with the addition of exogenous neutrophils. Samples were collected at five specific time points: Day 1, Day 4, Day 8, Day 11, and Day 14, to monitor this progression. Cells from each time point were analyzed using CITE-seq (RNA + ADT). 
The alignment and quantification pipeline was not performed as part of this work. This analysis starts directly from the gene count matrices generated by 10X Genomics.
Specifically, the input data consist of the filtered_feature_bc_matrix.h5 and raw_feature_bc_matrix.h5 files, which correspond to the processed outputs of the Cell Ranger pipeline (10X Genomics).

## Resources 
R and RStudio (version ≥ 4.3.x)
R packages, primarily:
- Seurat v5.2.0
* Additional dependencies listed in the renv environment (see the env folder)

The project uses renv to manage R dependencies, ensuring reproducibility. The environment is compatible with R version ≥ 4.3.x  

Python (used for complementary tools, details available in the relevant scripts)

## Author 
Augustine Blanc-Boekholt 
