---
title: "infer_CNA"
params: 
  patient: "P2"
  patientbis: "P3"
  file: "~/Documents/Singlecell-R/Scripts/Matrix/CLL_Seurat_P2"
  filebis: "~/Documents/Singlecell-R/Scripts/Matrix/CLL_Seurat_P3"
format: html
editor: visual
---

## Copy Number Alteration calculation

Goal : track the cancer cell thank to the CNA

```{r}
library(Seurat)
library(copykat)
library(ggplot2)
library(stringr)
library(numbat)
```

```{r}
data_p2<-readRDS(params$file)
data_p3<-readRDS(params$filebis)
```

## CopyKat 

Run copyKat (timepoint separately)

Patient 2

```{r}
timepoints <- c("D1", "D4", "D8", "D11", "D14")
for (tp in timepoints) {

  data_tp <- subset(data_p2, subset = timepoint == tp)
  counts_matrix <- GetAssayData(data_tp, assay = "RNA", layer = "counts")
  norm.cells <- colnames(data_tp)[data_tp$Annotation %in% c("Monocyte", "CD4 T", "CD8 T", "NK")]
  
  ck <- copykat(
    rawmat = counts_matrix,
    id.type = "S",
    ngene.chr = 5,
    win.size = 25,
    KS.cut = 0.1,
    sam.name = tp,
    distance = "euclidean",
    norm.cell.names = norm.cells,
    output.seg = "FALSE",
    plot.genes = "TRUE",
    genome = "hg20",
    n.cores = 1
  )
  
  copykat_results[[tp]] <- ck
  pred_results[[tp]] <- data.frame(ck$prediction)
  cna_results[[tp]] <- data.frame(ck$CNAmat)
}
```

Patient 3

```{r}
timepoints <- c("D1", "D4", "D8", "D11", "D14")
for (tp in timepoints) {

  data_tp <- subset(data_p3, subset = timepoint == tp)
  counts_matrix <- GetAssayData(data_tp, assay = "RNA", layer = "counts")
  norm.cells <- colnames(data_tp)[data_tp$Annotation %in% c("Monocyte", "NK")]
  
  ck <- copykat(
    rawmat = counts_matrix,
    id.type = "S",
    ngene.chr = 5,
    win.size = 25,
    KS.cut = 0.1,
    sam.name = tp,
    distance = "euclidean",
    norm.cell.names = norm.cells,
    output.seg = "FALSE",
    plot.genes = "TRUE",
    genome = "hg20",
    n.cores = 1
  )
  
  copykat_results[[tp]] <- ck
  pred_results[[tp]] <- data.frame(ck$prediction)
  cna_results[[tp]] <- data.frame(ck$CNAmat)
}
```

Exploring the results

Add the results in the Seurat Object, day sep

```{r}
pred_D1 <- data.frame(copykat_D1$prediction)
data$copykat_pred <- NA
rownames(pred.res_D1) <- pred_D1$cell.names
data$copykat_pred <- pred_D1[colnames(data), "copykat.pred"]
```

```{r}
DimPlot(data, group.by = "copykat_pred", reduction = "umap", pt.size=0.1, label.size = 2.5) + ggtitle(str_wrap("UMAP", width = 60)) + theme(plot.title = element_text(size = 15, hjust = 0.5))

```

Timepoints combined

Patient 2

```{r}
copykat.res <- do.call(rbind, pred_results)data_p2$copykat_pred <- NA
rownames(copykat.res) <- copykat.res$cell.names
data_p2$copykat_pred <- copykat.res[colnames(data_p2), "copykat.pred"]

title=paste("UMAP CNA, patient", params$patient)
plot<-DimPlot(data_p2, group.by = "copykat_pred", reduction = "umap", pt.size=0.05, label.size = 2.5) + ggtitle(str_wrap(title, width = 60)) + theme(plot.title = element_text(size = 10, hjust = 0.5))
ggsave(paste0("~/Documents/Singlecell-R/Scripts/Plots/cna_", params$patient, ".jpeg"), plot, width = 3000, height = 2000, units = "px")
```

Patient 3

```{r}
copykat.res <- do.call(rbind, pred_results)data_p3$copykat_pred <- NA
rownames(copykat.res) <- copykat.res$cell.names
data_p3$copykat_pred <- copykat.res[colnames(data_p3), "copykat.pred"]

title=paste("UMAP CNA, patient", params$patientbis)
plot<-DimPlot(data_p3, group.by = "copykat_pred", reduction = "umap", pt.size=0.05, label.size = 2.5) + ggtitle(str_wrap(title, width = 60)) + theme(plot.title = element_text(size = 10, hjust = 0.5))
ggsave(paste0("~/Documents/Singlecell-R/Scripts/Plots/cna_", params$patientbis, ".jpeg"), plot, width = 3000, height = 2000, units = "px")
```

## Numbat

Patient 2

Barcodes files

```{r}
timepoints <- c("D1", "D4", "D8", "D11", "D14")
for (tp in timepoints) {
  data_sub<-subset(data_p2, subset = timepoint == tp)
  counts_matrix <- GetAssayData(data_sub, assay = "RNA", layer = "counts")
  barcodes <- colnames(counts_matrix)
  barcodes <- sub("^D[0-9]+_", "", barcodes)
  write.table(barcodes, paste0("~/Documents/Singlecell-R/Scripts/Data/barcodes_", tp, ".tsv"), sep="\t",row.names = FALSE, col.names = FALSE)
}
```

Run the pile up and phasing on Numbat Docker in the HPC : run the script Numbat.sh

Count matrix

```{r}
counts_matrix <- GetAssayData(data_p2, assay = "RNA", layer = "counts")
colnames(counts_matrix) <- sub("^D[0-9]+_", "", colnames(counts_matrix))
```

Reference created (or choose the given one)

```{r}
data_ref<-subset(data_p2, subset = (Annotation %in% c("CD4 T", "CD8 T", "NK", "Monocyte")))
cell_annot <- data.frame(
  cell = sub("^D[0-9]+_", "", colnames(data_ref)),
  group = data_ref$Annotation,
  stringsAsFactors = FALSE
)
counts_ref <- GetAssayData(data_ref, assay = "RNA", layer = "counts")
colnames(counts_ref) <- sub("^D[0-9]+_", "", colnames(counts_ref))

ref_internal <- aggregate_counts(counts_ref, cell_annot)
```

Running Numbat

```{r}
df_allele<-read.table("~/Documents/Singlecell-R/Scripts/Matrix/D1_allele_counts.tsv", sep="\t", header=T)
```

```{r}
out = run_numbat(
    counts_matrix, # gene x cell integer UMI count matrix 
    ref_hca, # reference expression profile, a gene x cell type normalized expression level matrix
    df_allele, # allele dataframe generated by pileup_and_phase script
    genome = "hg38",
    t = 1e-5,
    ncores = 4,
    plot = TRUE,
    out_dir = './test'
)
```
