---
title: "Timepoint analysis and trajectory"
params: 
  patient: "P2"
  file: "Matrix/CLL_annotated_Seurat_LogNorm_P2.rds"
  type: "DDQC"
format: html
editor: visual
---

## Timepoint analysis

The analyzed dataset consists of five timepoints. The goal of timepoint analysis in single-cell data is to reconstruct the dynamic processes that cells undergo, such as differentiation or activation. In our study, we aim to determine whether specific clusters or transitional states, particularly the bridge between B cells and macrophages and T cells, emerge or evolve over the 14-day time course. This approach allows us to capture temporal changes in cell states.

```{r}
#| message: false
library(SeuratObject)
library(sp)
library(Seurat)
library(hdf5r)
library(dplyr)
library(patchwork)
library(ggplot2)
library(SeuratData)
library(pheatmap)
library(viridis)
library(stringr)
library(monocle3)
library(scCustomize)
library(slingshot)
library(SeuratWrappers)
library(BPCells)
library(CytoTRACE2) 
library(SeuratDisk)
source("functions.R")
```

```{r}
data<-readRDS(params$file)
```

## Complete data

UMAP by time

```{r}
data$timepoint <- factor(
  data$timepoint,
  levels = c("D1", "D4", "D8", "D11", "D14")
)
title=paste("UMAP timepoint, pre-processing type :", params$type, ", patient", params$patient)
plotumap<-DimPlot(data, group.by = "timepoint", reduction = "umap", pt.size=0.1, label.size = 2.5) + ggtitle(str_wrap(title, width = 60)) + theme(plot.title = element_text(size = 12, hjust = 0.5))
plotumap
ggsave(paste0(params$type,"_days_umap_", params$patient, ".jpeg"), plotumap, width = 3000, height = 2000, units = "px")
```

UMAP by timepoint

```{r}
data$timepoint <- factor(
  data$timepoint,
  levels = c("D1", "D4", "D8", "D11", "D14")
)

cell_palette <- c(
  "CD4 T" = "#1f77b4",
  "CD8 T" = "#2ca02c",
  "NK" = "#e7298a",
  "B naive" = "#ff7f0e",
  "B intermediate" = "#ffd92f",
  "B memory" = "#582d15ff",
  "Monocyte" = "#e41a1c",
  "Macrophage" = "#984ea3",
  "DC" = "#4daf4a",
  "Neutrophil" = "#377eb8",
  "Platelet" = "#999999",
  "Erythroblast" = "#17becf",
  "Plasmablast" = "#fdbf6f",
  "HSPC" = "#6a3d9a",
  "B" = "#f2a890ff",
  "T" = "#93f290ff")

title=paste("UMAP by timepoint, pre-processing type :", params$type, ", patient", params$patient)
plotumap<-DimPlot(data, reduction = "umap", group.by = "final_annot", split.by = "timepoint", cols = cell_palette) + ggtitle(str_wrap(title, width = 100)) + theme(plot.title = element_text(size = 13, hjust = 0.5))
plotumap
ggsave(paste0(params$type,"_timepoint_umap_", params$patient, ".jpeg"), plotumap, width = 5000,
  height = 2000, units = "px")
```

## Focus on B cells + macrophages

```{r}
data_Bxmacro<-subset(data, subset = final_annot=="B"|final_annot=="B intermediate"| final_annot=="B memory"|final_annot=="B naive"|final_annot=="Macrophage")
```

UMAP by time

```{r}
data_Bxmacro$timepoint <- factor(
  data_Bxmacro$timepoint,
  levels = c("D1", "D4", "D8", "D11", "D14")
)
title=paste("UMAP timepoint, only B cells and macrophages, pre-processing type :", params$type, ", patient", params$patient)
plotumap<-DimPlot(data_Bxmacro, group.by = "timepoint", reduction = "umap", pt.size=0.1, label.size = 2.5) + ggtitle(str_wrap(title, width = 60)) + theme(plot.title = element_text(size = 15, hjust = 0.5))
plotumap
ggsave(paste0(params$type,"_bxmacr_days_umap_", params$patient, ".jpeg"), plotumap, width = 3000, height = 2000, units = "px")
```

UMAP by timepoint

```{r}
data_Bxmacro$timepoint <- factor(
  data_Bxmacro$timepoint,
  levels = c("D1", "D4", "D8", "D11", "D14")
)
title=paste("UMAP by timepoint, only B cells and macrophages, pre-processing type :", params$type, ", patient", params$patient)
plotumap<-DimPlot(data_Bxmacro, reduction = "umap", group.by = "final_annot", split.by = "timepoint", cols = cell_palette) + ggtitle(str_wrap(title, width = 100)) + theme(plot.title = element_text(size = 15, hjust = 0.5))
plotumap
ggsave(paste0(params$type,"_bxmacr_timepoint_umap_", params$patient, ".jpeg"), plotumap, width = 5000,
  height = 2000, units = "px")
```

## Focus on the B cells + T cells

```{r}
data_BxT<-subset(data, subset = final_annot=="B"|final_annot=="B intermediate"| final_annot=="B memory"|final_annot=="B naive"|final_annot=="CD4 T"|final_annot=="CD8 T"|final_annot=="T")
```

UMAP by time

```{r}
data_BxT$timepoint <- factor(
  data_BxT$timepoint,
  levels = c("D1", "D4", "D8", "D11", "D14")
)
title=paste("Trajectory UMAP, only B cells and T cells, pre-processing type :", params$type, ", patient", params$patient)
plotumap<-DimPlot(data_BxT, group.by = "timepoint", reduction = "umap", pt.size=0.1, label.size = 2.5) + ggtitle(str_wrap(title, width = 60)) + theme(plot.title = element_text(size = 15, hjust = 0.5))
plotumap
ggsave(paste0(params$type,"_bxt_days_umap_", params$patient, ".jpeg"), plotumap, width = 3000, height = 2000, units = "px")
```

UMAP by timepoint

```{r}
data_BxT$timepoint <- factor(
  data_BxT$timepoint,
  levels = c("D1", "D4", "D8", "D11", "D14")
)
title=paste("UMAP by timepoint, only B cells and T cells, pre-processing type :", params$type, ", patient", params$patient)
plotumap<-DimPlot(data_BxT, reduction = "umap", group.by = "final_annot", split.by = "timepoint", cols = cell_palette) + ggtitle(str_wrap(title, width = 100)) + theme(plot.title = element_text(size = 15, hjust = 0.5))
plotumap
ggsave(paste0(params$type,"_bxt_timepoint_umap_", params$patient, ".jpeg"), plotumap, width = 5000,
  height = 2000, units = "px")
```

T cells

```{r}
data_T<-subset(data, subset = final_annot=="CD4 T"|final_annot=="CD8 T"|final_annot=="T")
data_T$timepoint <- factor(
  data_T$timepoint,
  levels = c("D1", "D4", "D8", "D11", "D14")
)
title=paste("Trajectory UMAP, only T cells, pre-processing type :", params$type, ", patient", params$patient)
plotumap<-DimPlot(data_T, group.by = "timepoint", reduction = "umap", pt.size=0.1, label.size = 2.5) + ggtitle(str_wrap(title, width = 60)) + theme(plot.title = element_text(size = 15, hjust = 0.5))
plotumap
ggsave(paste0(params$type,"_t_days_umap_", params$patient, ".jpeg"), plotumap, width = 3000, height = 2000, units = "px")
```

B cells

```{r}
data_B<-subset(data, subset = final_annot=="B"|final_annot=="B intermediate"| final_annot=="B memory"|final_annot=="B naive")
data_B$timepoint <- factor(
  data_B$timepoint,
  levels = c("D1", "D4", "D8", "D11", "D14")
)
title=paste("Trajectory UMAP, only B cells, pre-processing type :", params$type, ", patient", params$patient)
plotumap<-DimPlot(data_B, group.by = "timepoint", reduction = "umap", pt.size=0.1, label.size = 2.5) + ggtitle(str_wrap(title, width = 60)) + theme(plot.title = element_text(size = 15, hjust = 0.5))
plotumap
ggsave(paste0(params$type,"_b_days_umap_", params$patient, ".jpeg"), plotumap, width = 3000, height = 2000, units = "px")
```

Macrophages

```{r}
data_macro<-subset(data, subset = final_annot=="Macrophage")
data_macro$timepoint <- factor(
  data_macro$timepoint,
  levels = c("D1", "D4", "D8", "D11", "D14")
)
title=paste("Trajectory UMAP, only B cells and T cells, pre-processing type :", params$type, ", patient", params$patient)
plotumap<-DimPlot(data_macro, group.by = "timepoint", reduction = "umap", pt.size=0.1, label.size = 2.5) + ggtitle(str_wrap(title, width = 60)) + theme(plot.title = element_text(size = 15, hjust = 0.5))
plotumap
ggsave(paste0(params$type,"_macro_days_umap_", params$patient, ".jpeg"), plotumap, width = 3000, height = 2000, units = "px")
```

## Trajectory

We used Monocle 3 to reconstruct cellular trajectories and infer the pseudotime ordering of cells based on their gene expression profiles. This approach allows us to model dynamic biological processes, such as differentiation or activation, by arranging cells along a trajectory that reflects their transcriptional progression.

Trajectory on days separately macrophages & B cells

```{r}
timepoints <- c("D1", "D4", "D8", "D11", "D14")
plot_list <- list()

# 1. Compute the maximum pseudotime scale
max_pt <- 0
for(tp in timepoints){
  data_sub <- subset(data, subset = timepoint == tp & 
                      (final_annot %in% c("Macrophage", "B", "B intermediate", 
                                          "B memory", "B naive")))
  cds <- as.cell_data_set(data_sub)
  cds@clusters$UMAP$clusters <- data_sub@active.ident
  partition <- as.factor(rep(1, ncol(data_sub)))
  names(partition) <- colnames(data_sub)
  cds@clusters$UMAP$partitions <- partition
  cds@int_colData@listData$reducedDims$UMAP <- data_sub@reductions$umap@cell.embeddings
  colData(cds)$cell.type <- data_sub$final_annot
  cds <- learn_graph(cds, use_partition = FALSE)
  cds <- order_cells(cds, root_cells = colnames(data_sub)[data_sub$final_annot == "B naive"])
  
  max_pt <- max(max_pt, max(pseudotime(cds), na.rm = TRUE))
}

# 2. Generate the plots without legends
for(tp in timepoints){
  data_sub <- subset(data, subset = timepoint == tp & 
                      (final_annot %in% c("Macrophage", "B", "B intermediate", 
                                          "B memory", "B naive")))
  cds <- as.cell_data_set(data_sub)
  cds@clusters$UMAP$clusters <- data_sub@active.ident
  partition <- as.factor(rep(1, ncol(data_sub)))
  names(partition) <- colnames(data_sub)
  cds@clusters$UMAP$partitions <- partition
  cds@int_colData@listData$reducedDims$UMAP <- data_sub@reductions$umap@cell.embeddings
  colData(cds)$cell.type <- data_sub$final_annot
  cds <- learn_graph(cds, use_partition = FALSE)
  cds <- order_cells(cds, root_cells = colnames(data_sub)[data_sub$final_annot == "B naive"])
  
  p <- plot_cells(
    cds,
    color_cells_by = "pseudotime",
    show_trajectory_graph = TRUE,
    label_groups_by_cluster = FALSE,
    label_leaves = FALSE,
    label_branch_points = FALSE, 
    graph_label_size = 0
  ) + 
    scale_color_viridis_c(limits = c(0, max_pt), guide = "none") + # remove legend
    theme_void() + 
    ggtitle(tp)
  
  plot_list[[tp]] <- p
}

# 3. Combine the plots and add a common legend
# To get a single legend, create an empty plot with the legend
legend_plot <- ggplot(data.frame(x = 1, y = 1, pseudotime = c(0, max_pt)), 
                      aes(x, y, color = pseudotime)) +
  geom_point() +
  scale_color_viridis_c(limits = c(0, max_pt)) +
  guides(color = guide_colorbar(title = "Pseudotime")) +
  theme_void() + theme(legend.position = "right")

# Extract the legend
library(cowplot)
legend <- get_legend(legend_plot)

# Assemble the 5 plots side by side + single legend
final_plot <- plot_grid(plotlist = plot_list, ncol = 5)
final_plot_with_legend <- plot_grid(final_plot, legend, rel_widths = c(5, 1))

# Add a general title
title <- ggdraw() + draw_label(paste("Trajectory with Monocle3 pseudotime B cells and Macrophages, patient :",params$patient), fontface='bold')
final_plot_with_title <- plot_grid(title, final_plot_with_legend, ncol = 1, rel_heights = c(0.05, 1))

# Display
final_plot_with_title

ggsave(paste0(params$type,"_traj_", params$patient, ".jpeg"), final_plot_with_title, width = 4000,
  height = 1500, units = "px")
```

```{r}
timepoints <- c("D1", "D4", "D8", "D11", "D14")
plot_list <- list()

# 1. Compute the maximum pseudotime scale
max_pt <- 0
for(tp in timepoints){
  data_sub <- subset(data, subset = timepoint == tp & 
                      (final_annot %in% c("Macrophage", "B", "B intermediate", 
                                          "B memory", "B naive", "T", "CD4 T", "CD8 T", "Monocyte", "NK")))
  cds <- as.cell_data_set(data_sub)
  cds@clusters$UMAP$clusters <- data_sub@active.ident
  partition <- as.factor(rep(1, ncol(data_sub)))
  names(partition) <- colnames(data_sub)
  cds@clusters$UMAP$partitions <- partition
  cds@int_colData@listData$reducedDims$UMAP <- data_sub@reductions$umap@cell.embeddings
  colData(cds)$cell.type <- data_sub$final_annot
  cds <- learn_graph(cds, use_partition = FALSE)
  cds <- order_cells(cds, root_cells = colnames(data_sub)[data_sub$final_annot == "B naive"])
  
  max_pt <- max(max_pt, max(pseudotime(cds), na.rm = TRUE))
}

# 2. Generate the plots without legends
for(tp in timepoints){
  data_sub <- subset(data, subset = timepoint == tp & 
                      (final_annot %in% c("Macrophage", "B", "B intermediate", 
                                          "B memory", "B naive", "T", "CD4 T", "CD8 T", "Monocyte", "NK")))
  cds <- as.cell_data_set(data_sub)
  cds@clusters$UMAP$clusters <- data_sub@active.ident
  partition <- as.factor(rep(1, ncol(data_sub)))
  names(partition) <- colnames(data_sub)
  cds@clusters$UMAP$partitions <- partition
  cds@int_colData@listData$reducedDims$UMAP <- data_sub@reductions$umap@cell.embeddings
  colData(cds)$cell.type <- data_sub$final_annot
  cds <- learn_graph(cds, use_partition = FALSE)
  cds <- order_cells(cds, root_cells = colnames(data_sub)[data_sub$final_annot == "B naive"])
  
  p <- plot_cells(
    cds,
    color_cells_by = "pseudotime",
    show_trajectory_graph = TRUE,
    label_groups_by_cluster = FALSE,
    label_leaves = FALSE,
    label_branch_points = FALSE, 
    graph_label_size = 0
  ) + 
    scale_color_viridis_c(limits = c(0, max_pt), guide = "none") + # remove legend
    theme_void() + 
    ggtitle(tp)
  
  plot_list[[tp]] <- p
}

# 3. Combine the plots and add a common legend
# To get a single legend, create an empty plot with the legend
legend_plot <- ggplot(data.frame(x = 1, y = 1, pseudotime = c(0, max_pt)), 
                      aes(x, y, color = pseudotime)) +
  geom_point() +
  scale_color_viridis_c(limits = c(0, max_pt)) +
  guides(color = guide_colorbar(title = "Pseudotime")) +
  theme_void() + theme(legend.position = "right")

# Extract the legend
library(cowplot)
legend <- get_legend(legend_plot)

# Assemble the 5 plots side by side + single legend
final_plot <- plot_grid(plotlist = plot_list, ncol = 5)
final_plot_with_legend <- plot_grid(final_plot, legend, rel_widths = c(5, 1))

# Add a general title
title <- ggdraw() + draw_label(paste("Trajectory with Monocle3 pseudotime B, T cells, NK, Monocytes and Macrophages, patient :",params$patient), fontface='bold')
final_plot_with_title <- plot_grid(title, final_plot_with_legend, ncol = 1, rel_heights = c(0.05, 1))

# Display
final_plot_with_title

ggsave(paste0(params$type,"_traj_", params$patient, ".jpeg"), final_plot_with_title, width = 4000,
  height = 1500, units = "px")
```

D8

```{r}
data_sub <- subset(data, subset = timepoint == "D8" & (final_annot %in% c("Macrophage", "B", "B intermediate", 
                                          "B memory", "B naive", "T", "CD4 T", "CD8 T", "Monocyte", "NK")))
cds <- as.cell_data_set(data_sub)
cds@clusters$UMAP$clusters <- data_sub@active.ident
partition <- as.factor(rep(1, ncol(data_sub)))
names(partition) <- colnames(data_sub)
cds@clusters$UMAP$partitions <- partition
cds@int_colData@listData$reducedDims$UMAP <- data_sub@reductions$umap@cell.embeddings
colData(cds)$cell.type <- data_sub$final_annot
cds <- learn_graph(cds, use_partition = FALSE)
cds <- order_cells(cds, root_cells = colnames(data_sub)[data_sub$final_annot == "B naive"])

p1<-plot_cells(
    cds,
    color_cells_by = "pseudotime",
    show_trajectory_graph = TRUE,
    label_groups_by_cluster = FALSE,
    label_leaves = FALSE,
    label_branch_points = FALSE, 
    graph_label_size = 0
  ) + 
    ggtitle("D8")
```

D11

```{r}
data_sub <- subset(data, subset = timepoint == "D11" & (final_annot %in% c("Macrophage", "B", "B intermediate", 
                                          "B memory", "B naive", "T", "CD4 T", "CD8 T", "Monocyte", "NK")))
cds <- as.cell_data_set(data_sub)
cds@clusters$UMAP$clusters <- data_sub@active.ident
partition <- as.factor(rep(1, ncol(data_sub)))
names(partition) <- colnames(data_sub)
cds@clusters$UMAP$partitions <- partition
cds@int_colData@listData$reducedDims$UMAP <- data_sub@reductions$umap@cell.embeddings
colData(cds)$cell.type <- data_sub$final_annot
cds <- learn_graph(cds, use_partition = FALSE)
cds <- order_cells(cds, root_cells = colnames(data_sub)[data_sub$final_annot == "B naive"])
  
p2<-plot_cells(
    cds,
    color_cells_by = "pseudotime",
    show_trajectory_graph = TRUE,
    label_groups_by_cluster = FALSE,
    label_leaves = FALSE,
    label_branch_points = FALSE, 
    graph_label_size = 0
  ) + 
    ggtitle("D11")
```

D8+D11

```{r}
data_sub <- subset(data, subset = timepoint %in% c("D8", "D11") & (final_annot %in% c("Macrophage", "B", "B intermediate", "B memory", "B naive", "T", "CD4 T", "CD8 T", "Monocyte", "NK")))

cds <- as.cell_data_set(data_sub)
cds@clusters$UMAP$clusters <- data_sub@active.ident
partition <- as.factor(rep(1, ncol(data_sub)))
names(partition) <- colnames(data_sub)
cds@clusters$UMAP$partitions <- partition
cds@int_colData@listData$reducedDims$UMAP <- data_sub@reductions$umap@cell.embeddings
colData(cds)$cell.type <- data_sub$final_annot
cds <- learn_graph(cds, use_partition = FALSE)
cds <- order_cells(cds, root_cells = colnames(data_sub)[data_sub$final_annot == "B naive"])
  
p3<-plot_cells(
    cds,
    color_cells_by = "pseudotime",
    show_trajectory_graph = TRUE,
    label_groups_by_cluster = FALSE,
    label_leaves = FALSE,
    label_branch_points = FALSE, 
    graph_label_size = 0
  ) + 
    ggtitle("D8+D11")

plot<-p1+p2+p3
ggsave(paste0("Plots/traj_d8_d11_", params$patient, ".jpeg"), plot, width = 4000,
  height = 1500, units = "px")
```

Trajectory with Slingshot

```{r}
data_sub <- subset(data, subset = timepoint == "D8" & 
                      (final_annot %in% c("Macrophage", "B", "B intermediate", 
                                          "B memory", "B naive", "T", "CD4 T", "CD8 T", "Monocyte", "NK")))

dimred <- data_sub@reductions$umap@cell.embeddings
clustering <- factor(data_sub$final_annot)
counts <- as.matrix(GetAssayData(data_sub, assay = "RNA", slot = "counts")[VariableFeatures(data_sub), ])

lineages <- getLineages(data = dimred, clusterLabels = clustering, start.clus = c("B","T", "Monocyte"), end.clus = c("Macrophage","NK"))

# Curves
curves <- getCurves(lineages, approx_points = 300, thresh = 0.01, stretch = 0.8, allow.breaks = FALSE, shrink = 0.99)

pt <- slingPseudotime(curves)[,1]
col_pt <- viridis::viridis(100)[cut(pt, breaks = 100)]

pdf("slingshot_pseudotime.pdf", width = 7, height = 6)
plot<-plot(dimred, col = col_pt, asp = 1, pch = 16)
lines(SlingshotDataSet(curves), lwd = 3, col = "black")
dev.off()

ggsave(paste0("Plots/slingshot_d8_d11_", params$patient, ".jpeg"), plot, width = 4000,
  height = 1500, units = "px")
```

Trajectory with STREAM2 in Python

```{r}
obj <- Convert_Assay(seurat_object = data_sub, assay='RNA', convert_to = "V3")
obj@assays <- obj@assays["RNA"]
SaveH5Seurat(obj, filename = "Matrix/data_v4.h5Seurat", overwrite = TRUE)
Convert("Matrix/data_v4.h5Seurat", dest = "h5ad")
```

## Pluripotency

```{r}
data_bis<-subset(data, subset = final_annot=="B"|final_annot=="B intermediate"| final_annot=="B memory"|final_annot=="B naive"|final_annot=="Macrophage"& timepoint=="D1")
expression_data <-data_bis[[ "RNA" ]]$counts
cytotrace2_result <- cytotrace2(expression_data)
annotation <- data_bis$final_annot
plots <- plotData(cytotrace2_result = cytotrace2_result, 
                  annotation = annotation,
                  expression_data = expression_data
                  ) 
plots$name_of_the_plot
plot1<-plots$CytoTRACE2_UMAP
ggsave(paste0("cytotraceumap_d4_", params$patient, ".jpeg"), plot1, width = 3000,
  height = 1500, units = "px")
plot1<-plots$CytoTRACE2_Boxplot_byPheno
ggsave(paste0("cytotraceboxplot_d4_", params$patient, ".jpeg"), plot1, width = 3000,
  height = 1500, units = "px")
```
